proxy/src/coco.rs

```diff
@@ -88,20 +90,32 @@ impl Peer {
         clippy::wildcard_enum_match_arm,
         clippy::match_wildcard_for_single_variants
     )]
-    pub fn list_projects(&self) -> Result<Vec<project::Project<entity::Draft>>, error::Error> {
-        self.with_api(|api| {
-            let storage = api.storage();
-            Ok(storage
-                .all_metadata()?
-                .flat_map(|entity| {
-                    entity.ok()?.try_map(|info| match info {
-                        entity::data::EntityInfo::Project(info) => Some(info),
-                        _ => None,
+    pub fn list_projects(&self) -> Result<Vec<Project>, error::Error> {
+        let api = self.api.lock().map_err(|_| error::Error::LibradLock)?;
+        let storage = api.storage();
+        let project_meta = storage.all_metadata()?.flat_map(|entity| {
+            entity.ok()?.try_map(|info| match info {
+                entity::data::EntityInfo::Project(info) => Some(info),
+                _ => None,
+            })
+        });
+        project_meta
+            .map(|project| {
+                api.storage()
+                    .open_repo(project.urn())
+                    .map_err(error::Error::from)
+                    .and_then(|repo| {
+                        repo.browser(project.default_branch())
+                            .map_err(error::Error::from)
+                            .and_then(|browser| {
+                                browser
+                                    .get_stats()
+                                    .map_err(error::Error::from)
+                                    .map(|stats| Project::from_project_stats(project, stats))
+                            })
```

Comment:
I can see @FintanH style all over it. While it's smart, it's a bit dense and hard to comprehend, maybe we can break it up in smaller bits.

Reply:
It's not my fault I can't use `?` in fucking `map` :weary: I was saying to @sarahscott that this is showing a code smell around my `with_*` code because I was running into deadlocks.

Reply:
I think I have an idea to clean this up and avoid deadlocks